### DISCUSSION TITLE
Handling conflicts w/ or w/o precedence, terminal token

### DISCUSSION THREAD

**Discussion by tmatijevich** [tree-sitter/tree-sitter#1052] [Q&A] [answered]
Hello and thank you for the tree-sitter project.

I am stuck on the part of my grammar involving a switch statement. I'm often getting generation conflicts or parsing errors, even with some precedence attempts. I've found some success with larger terminal tokens - though I'm not sure if that is best practice as I scale the language.

Here is a mwe of my grammar with the case statement. I think the difficulty is a `case_value` can have an `identifier` followed by a `':'` while an `assignment` begins with an `identifier` followed by `':='`. This seems ambiguous.

```js
module.exports = grammar({
  name: 'structured_text',
  
  word: $ => $.identifier,
  
  rules: {
    source_file: $ => repeat($.statement),
    
    statement: $ => choice(
      $.assignment,
      $.case_statement,
    ),
    
    case_statement: $ => seq(
      'CASE',
      field('CaseControlValue', $.identifier),
      'OF',
      repeat($.case),
      optional($.else_case),
      'END_CASE',
      optional(';')
    ),
    
    case: $ => seq(
      $.case_value,
      repeat($.statement)
    ),
    
    else_case: $ => seq(
      'ELSE',
      repeat($.statement)
    ),
    
    case_value: $ => {
      const num = /\d+/;
      const range = seq(num, '..', num);
      const varConst = /[a-zA-Z_]\w*/;
      return token(seq(choice(num, range, varConst), ':'));
    },
    
    assignment: $ => seq(
      $.identifier,
      ':=',
      $.number,
      ';'
    ),
    
    identifier: $ => token(/[a-zA-Z_]\w*/),
    
    number: $ => token(/\d+/)
  }
});
```

I've only had success when placing `':'` inside the `case_value` token to successfully distinguish between the start of a `case_value` versus the start of an `assignment`.

- However, this doesn't allow for space between `varConst` and `':'` (as expected from the token).
  - `return token(seq(choice(num, range, varConst), '/\s*:/'));` then breaks assignments in the parsing sample

Here is my sample. I've tried numbered precedencies, `prec.right`, and the dynamic precedence with the conflicts array - but I can't seem to get the `:` versus `:=` to budge.

```
Var1 := 25;

CASE State OF

    0: 
        x := 5;
        y := 5;
        
    1 :
        z := 6;
        Var1 := 100;
        
    Case3:
    
    ELSE
        Test := 6458;
        
END_CASE
```

Is there a recommendation for addressing this `case_value` versus `assignment` ambiguity? Either with (dynamic) precedence or improved placement of tokens?

**Comment by alemuller** [comment, marked as answer]
Your parser seems to break with `token(...'/\s*:/')` because `case_value` and `identifier` are [conflicting tokens](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens).

When `case_value` matches `\s*:`, there are two possible tokens:
```
foo := x
^^^ identifier "foo" length=3
^^^^ case_value "foo :" length=5
```

And, according to the documentation, the longer token "foo :" (`case_value`) should be matched instead of "foo" (`identifier`). 

But there isn't any rule for "`seq( $.case_value, '=', ... )`" so the code `foo := ...` produces an `ERROR`.

---
Tokens are usually identifier, literals, reserved words and delimiters. It is not usual for not delimited literals to accept white spaces.

I would implement `$.case_value` using the `$.number`, `$.identifier` and a new rule for the range:

```js
module.exports = grammar({
  // ...
  rules: {
    // ...
    case_value: $ => seq(
      choice(
        $.number,
        $.range,
        $.identifier
      ),
      ':'
    ),

    range: $ => seq(
      field('lo',$.number),
      '..',
      field('hi',$.number),
    ),
    // ...
  }
});
```

Trying to generate the parser will produce an error:
```
$ tree-sitter generate

Unresolved conflict for symbol sequence:

  'CASE'  identifier  'OF'  case_value  •  identifier  …

Possible interpretations:

  1:  'CASE'  identifier  'OF'  (case  case_value  •  source_file_repeat1)
  2:  'CASE'  identifier  'OF'  (case  case_value)  •  identifier  …

Possible resolutions:

  1:  Specify a left or right associativity in `case`
  2:  Add a conflict for these rules: `case`
```

The conflict can be resolved following 2:

```js
module.exports = grammar({
  // ...
  conflicts: $ => [ [$.case] ],

  rules: {
    // ...
  }
});
```

**Reply by tmatijevich** [reply]
Ah, fantastic!

I was confused about the conflicts array (I thought it was supposed to be used in tandem with `prec.dynamic`).
But I see it's also used for the GLR - so I can explore multiple branches. Which is exact what I'm looking for to distinguish `identifier > ':'` versus `identifier > ':='` with enough "look-ahead".

Thank you @alemuller for looking at this.

**Reply by alemuller** [reply]
That's correct.

The `prec.dynamic` can be used with `conflicts` but it is not needed when there isn't any actual ambiguity.

I edited the post after you accept as an answer to explain why the code was breaking with `\s`.
